river/LayoutManager.zig

```diff
@@ -0,0 +1,356 @@
+            views,
+            self.output.main_count,
+            wl.Fixed.fromDouble(self.output.main_factor),
+            self.output.usable_box.width,
+            self.output.usable_box.height,
+            self.output.layout_demand.?.serial,
+        );
+
+        // And finally we advertise all visible views
+        var it = ViewStack(View).iter(self.output.views.first, .forward, self.output.pending.tags, Output.arrangeFilter);
+        while (it.next()) |view| {
+            self.layout.sendAdvertiseView(
+                view.pending.tags,
+                view.getAppId() orelse "[no-appid]",
+                self.output.layout_demand.?.serial,
+            );
+        }
+        self.layout.sendAdvertiseDone(self.output.layout_demand.?.serial);
+    }
+
+    fn handleRequest(layout: *zriver.LayoutV1, request: zriver.LayoutV1.Request, self: *Layout) void {
+        switch (request) {
+            .destroy => {
+                self.deinit();
+
+                // Remove this layout from the list
+                const node = @fieldParentPtr(std.TailQueue(Layout).Node, "data", self);
+                self.output.layouts.remove(node);
+                util.gpa.destroy(node);
+            },
+
+            // We receive this event when the client wants to push a view dimension proposal
+            // to the layout demand matching the serial.
+            .push_view_dimensions => |req| {
+                // Are we a valid layout?
+                if (self.namespace == null) {
+                    layout.postError(
+                        .no_namespace,
+                        "after receiving the namespace_in_use event, all requests other than destroy are illegal",
+                    );
+                    return;
+                }
+
+                // Does the serial match?
+                // We can't raise a protocol error when the serial is wrong, because we
+                // destroy the LayoutDemand after a timeout or a new layout demand request
+                // is raised, so that the client then has no idea that its previously
+                // perfectly fine serial is now invalid. So we just ignore requests with
+                // wrong serials.
+                if (self.output.layout_demand == null) return;
+                if (self.output.layout_demand.?.serial != req.serial) return;
+
+                self.output.layout_demand.?.pushViewDimensions(req.x, req.y, req.width, req.height) catch {};
+            },
+
+            // We receive this event when the client wants to mark the proposed layout
+            // of the layout demand matching the serial as done.
+            .commit => |req| {
+                // Are we a valid layout?
+                if (self.namespace == null) {
+                    layout.postError(
+                        .no_namespace,
+                        "after receiving the namespace_in_use event, all requests other than destroy are illegal",
+                    );
```

Comment:
This can't be a protocol error as it would be racy. We should just state that all requests other than destroy are ignored.